lib/ts/recipe/totp/types.ts

```diff
@@ -0,0 +1,168 @@
+            builder: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    issuer: string;
+    defaultSkew: number;
+    defaultPeriod: number;
+    getEmailOrPhoneForRecipeUserId?: GetEmailOrPhoneForRecipeUserIdFunc;
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type RecipeInterface = {
+    createDevice: (input: {
+        userId: string;
+        deviceName: string;
+        skew?: number;
+        period?: number;
+        userIdentifierInfo?: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK"; issuerName: string; secret: string; userIdentifier?: string; qrCode: string }
+        | { status: "DEVICE_ALREADY_EXISTS_ERROR" }
+    >;
+    verifyCode: (input: {
+        userId: string;
+        totp: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK" | "INVALID_TOTP_ERROR" }
+        | { status: "TOTP_NOT_ENABLED_ERROR" }
+        | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }
+    >;
+    verifyDevice: (input: {
+        userId: string;
+        deviceName: string;
+        totp: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK"; wasAlreadyVerified: boolean }
+        | { status: "TOTP_NOT_ENABLED_ERROR" }
+        | { status: "INVALID_TOTP_ERROR" | "UNKNOWN_DEVICE_ERROR" }
+        | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }
+    >;
+    updateDevice: (input: {
+        userId: string;
+        existingDeviceName: string;
+        newDeviceName: string;
+        userContext: any;
+    }) => Promise<{ status: "OK" | "TOTP_NOT_ENABLED_ERROR" | "DEVICE_ALREADY_EXISTS_ERROR" | "UNKNOWN_DEVICE_ERROR" }>;
+    removeDevice: (input: {
+        userId: string;
+        deviceName: string;
+        userContext: any;
+    }) => Promise<{ status: "OK"; didDeviceExist: boolean } | { status: "TOTP_NOT_ENABLED_ERROR" }>;
+    listDevices: (input: {
+        userId: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK"; devices: { name: string; period: number; skew: number; verified: boolean }[] }
+        | { status: "TOTP_NOT_ENABLED_ERROR" }
+    >;
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    appInfo: NormalisedAppinfo;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type APIInterface = {
+    createDevicePOST?: (input: {
+        session: SessionContainer;
+        deviceName: string;
```

Comment:
Tried MFA in AWS console?

Reply:
Answered on Slack. AWS Console has this.

Reply:
In this case, the input should be optional? How will the frontend know if it should ask for the device name or not?

Reply:
Okay. So we have two ways:
- Always use a default value for `deviceName` column (say "TOTP Device") unless frontend passes it while calling the API.  
- Our frontend SDK can check if any device exists. If yes, ask for new device name. Otherwise, use default value (say "TOTP Device")

Reply:
We can't rely on the frontend to always pass in a device name even if one device already exists. Therefore, you need to handle this case on the API as well - if a device already exists with "TOTP Device", then we need to use another name like "TOTP Device 2" or something.

Actually, we can just use "Device 1", "Device 2" etc by default.. unless the frontend passes a device name, then we use that. In case that device name already exists, we will return a `DEVICE_ALREADY_EXISTS_ERROR`

Reply:
Done.